text/0477-policy-validation.md

```diff
@@ -0,0 +1,739 @@
+to evaluate the rule set during development is still useful as it will
+improve detection speed and developer productivity.
+
+The goal of CDK Policy Validation is to minimize the amount of set up needed
+during development, and make it as easy as possible. For example, a typical
+workflow could be:
+
+* Developer writes a CDK application without the correct validations config.
+* Developer deploys non-compliant stacks.
+* Deployment guardrails catch these violations, and instruct the developer to add a validations property to their
+  application.
+
+#### Alternative
+
+```ts
+const app = new App();
+Policy.of(app).add(new CfnGuardValidator());
+```
+
+### Validation Report
+
+When you synthesize the CDK app the validator plugins will be called and the results will be printed.
+
+```text
+$ cdk synth
+
+Validation Report (CfnGuardValidator)
+-------------------------------------
+
+(Summary)
+
+╔═══════════╤════════════════════════╗
+║ Status    │ failure                ║
+╟───────────┼────────────────────────╢
+║ Plugin    │ CfnGuardValidator      ║
+╚═══════════╧════════════════════════╝
+
+(Violations)
+
+Ensure S3 Buckets are encrypted with a KMS CMK (1 occurrences)
+
+  Occurrences:
+  
+    - Construct Path: MyStack/MyCustomL3Construct/Bucket
+    - Stack Template Path: ./cdk.out/MyStack.template.json
+    - Creation Stack:
+        └──  MyStack (MyStack)
+             │ Library: aws-cdk-lib.Stack
+             │ Library Version: 2.50.0
+             │ Location: Object.<anonymous> (/home/hallcor/tmp/cdk-tmp-app/src/main.ts:25:20)
+             └──  MyCustomL3Construct (MyStack/MyCustomL3Construct)
+                  │ Library: N/A - (Local Construct)
+                  │ Library Version: N/A
+                  │ Location: new MyStack (/home/hallcor/tmp/cdk-tmp-app/src/main.ts:15:20)
+                  └──  Bucket (MyStack/MyCustomL3Construct/Bucket)
+                       │ Library: aws-cdk-lib/aws-s3.Bucket
+                       │ Library Version: 2.50.0
+                       │ Location: new MyCustomL3Construct (/home/hallcor/tmp/cdk-tmp-app/src/main.ts:9:20)
+    - Resource Name: my-bucket
+    - Locations:
+      > BucketEncryption/ServerSideEncryptionConfiguration/0/ServerSideEncryptionByDefault/SSEAlgorithm
+
+  Recommendation: Missing value for key `SSEAlgorithm` - must specify `aws:kms`
+  How to fix:
+    > Using override `app.findChild('my-bucket').addPropertyOverride('SSEAlgorithm', 'aws:kms');`
+    > Add to construct properties for `cdk-app/MyStack/Bucket`
+      `encryption: BucketEncryption.KMS`
+
+  
+Validation failed. See above reports for details
+```
+
+By default, the report will be printed in a human readable format. If you want a
+report in JSON format, enable it using the `@aws-cdk/core:validationReportJson`
+via the CLI or passing it directly to the application:
+
+```ts
+const app = new App({ 
+  context: { '@aws-cdk/core:validationReportJson': true }, 
```

Comment:
This is a bit strange. Normally JSON output is needed in order for a machine to parse the output, but running `cdk synth` also produces additional stuff that will make the output non parsable. That is, it will contains some lines, then a JSON report, and some more lines? 

I think its going to be hard to make sure the CLI doesn't print anything when this feature is turned on, so I would suggest allowing to pass a file where the JSON report will be written to. Is this possible via context?

Reply:
If the report returns anything that means it fails and synth will not write anything to `stdout`. The only thing written to `stdout` will be the report.